src/main/java/ch/njol/skript/expressions/ExprBookPages.java

```diff
@@ -54,38 +54,58 @@
 		Skript.registerExpression(ExprBookPages.class, String.class, ExpressionType.PROPERTY,
 				"[all] [the] [book] (pages|content) of %itemtypes%",
 				"%itemtypes%'s [book] (pages|content)",
-				"[book] page %number% of %itemtypes%", 
-				"%itemtypes%'s [book] page %number%");
+				"[book] page %number% of %itemtypes%",
+				"%itemtypes%'[s] [book] page %number%");
 	}
 	
-	private static final ItemType bookItem = Aliases.javaItemType("book with text");
-	
 	@SuppressWarnings("null")
-	private Expression<ItemType> book;
+	private Expression<ItemType> books;
 	@Nullable
-	private Expression<Number> page;
+	private Expression<Number> page = null;
+	
+	@SuppressWarnings({"unchecked", "null"})
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, ParseResult parseResult) {
+		books = (Expression<ItemType>) exprs[matchedPattern == 3 ? 1 : 0];
+		
+		if (matchedPattern > 1)
+			page = (Expression<Number>) exprs[matchedPattern ^ 2];
+		return true;
+	}
 	
 	@SuppressWarnings("null")
 	@Nullable
 	@Override
 	protected String[] get(Event e) {
-		ItemStack itemStack = book.getSingle(e).getRandom();
-		if (itemStack == null || !bookItem.isOfType(itemStack))
-			return null;
-		List<String> pages = ((BookMeta) itemStack.getItemMeta()).getPages();
-		if (page != null){
-			Number pageNumber = page.getSingle(e);
-			if (pageNumber == null){
-				return null;
-			}
-			int page = pageNumber.intValue();
-			if ((page) > pages.size() || page < 1){
-				return null;
+		ItemType[] items = books.getArray(e);
+		Number n = page != null ? this.page.getSingle(e) : null;
+		List<String> pages = new ArrayList<>();
+		
+		for (ItemType item : items) {
+			ItemMeta meta = item.getItemMeta();
+			
+			if (!(meta instanceof BookMeta))
+				continue;
+			
+			BookMeta book = (BookMeta) meta;
+			
+			if (!book.hasPages())
+				continue;
+			
+			if (n != null) {
+				int p = n.intValue();
+				if (p <= book.getPageCount())
+					pages.add(book.getPage(p));
+			} else {
+				for (int p = 1; p <= book.getPageCount(); p++)
```

Comment:
```
pages.addAll(book.getPages());
```

Reply:
I was looking for this method but missed it somehow